consensus/istanbul/uptime/autofix.go

```diff
@@ -0,0 +1,160 @@
+		// Provided header from the wrong epoch
+		return ErrWrongEpoch
+	}
+
+	lastHeader := af.builder.GetLastProcessedHeader()
+
+	if lastHeader == nil {
+		// It's from the same epoch and the builder is empty.
+		// Build up to the new header.
+		return af.cleanBuild(header)
+	}
+
+	lastHeaderNumber := lastHeader.Number.Uint64()
+
+	// If it's a rewind, rebuild
+	if number < lastHeaderNumber {
+		af.builder.Clear()
+		return af.cleanBuild(header)
+	}
+
+	// Normal sequential case
+	if number == lastHeaderNumber+1 {
+		if header.ParentHash == lastHeader.Hash() {
+			// Valid chain, add the header
+			return af.builder.ProcessHeader(header)
+		}
+		// Fork! Rebuild
+		af.builder.Clear()
+		return af.cleanBuild(header)
+	}
+
+	if number > lastHeaderNumber+1 {
+		// Normal advance (1 or more headers missing). Try to advance the builder
+		return af.advance(lastHeader, header)
+	}
+
+	// number == lastHeaderNumber
+	// check if it's an idempotent call or if
+	// there' a fork
+	if lastHeader.Hash() == header.Hash() {
+		// Nothing to do.
+		return nil
+	}
+
+	// fork, rebuild
+	af.builder.Clear()
+	return af.cleanBuild(header)
+}
+
+// cleanBuild does a clean build up to the header given, assuming the decorated builder is empty.
+func (af *autoFixBuilder) cleanBuild(upTo *types.Header) error {
+	epochSize := af.builder.GetEpochSize()
+	numberWithinEpoch := istanbul.GetNumberWithinEpoch(upTo.Number.Uint64(), epochSize)
+	// If this is the first header in the epoch, no need to call the provider
+	if numberWithinEpoch == 1 {
+		return af.builder.ProcessHeader(upTo)
+	}
+	// Else, load all the previous headers from the db.
+	headers, err := af.provider.GetEpochHeadersUpToLimit(epochSize, upTo, numberWithinEpoch)
+	if err != nil {
+		return err
+	}
+	return af.addAll(headers)
+}
+
+// advance advances the computation from the header given to the upTo header given.
+// This method requests to the provider the missing headers in between and adds them.
+// The only caveat is if the 'from' header is not in the same chain as the 'upTo' header.
+// In that case, a rebuild is made.
+func (af *autoFixBuilder) advance(from, upTo *types.Header) error {
+	upNumber := upTo.Number.Uint64()
+	fromNumber := from.Number.Uint64()
+	limit := upNumber - fromNumber + 1
+	headers, err := af.provider.GetEpochHeadersUpToLimit(af.builder.GetEpochSize(), upTo, limit)
+	if err != nil {
+		return err
+	}
+	// Verify that this is not a fork
+	if headers[0].Hash() != from.Hash() {
+		// Fork!
+		// Rebuild
+		af.builder.Clear()
+		return af.cleanBuild(upTo)
+	}
+	return af.addAll(headers[1:])
+}
+
+// addAll adds all the headers to the decorated builder.
+func (af *autoFixBuilder) addAll(headers []*types.Header) error {
+	for _, h := range headers {
+		err := af.builder.ProcessHeader(h)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (af *autoFixBuilder) ComputeUptime(epochLastHeader *types.Header) ([]*big.Int, error) {
+	err := af.ProcessHeader(epochLastHeader)
```

Comment:
The interface of ComputeUptime defines that it will return the uptime calculation, by sending the last header of the epoch. This implementation complies, so what we are talking about is actually implementation details, correct?

The calculation will not be necessarily corrupted because, if the autofix gets a ProcessHeader call from a different header, it will just recompute everything. The only possible problems I can think of are

1) ComputeUptime gets called too many times with different headers, therefore recalculating too much and it could be a performance problem. Is this even possible?

2) Someone is also using the decorated reference of the underlying builder, and we wish not modify that in this case. Again, I think this is not possible in the way we are using it, since it would be a violation of the standard use of a decoration pattern.

Reply:
I guess my point is that whenever the `miner` creates the `pending` block for the last block of an epoch, i will call `ComputeUptime` and within `ProcessHeader`. Or similarly, all validators generate blocks eagerly even if they are not the proposers. In both cases, you're "adding" a block to the block set of the epoch, which won't be mined, thus "corrupting" the computation.

So when the "actual" last block of epoch arrives, we will need to recompute the whole uptime. And this recomputation was a emergency recovery mechanism and not something we want to be happening all the time

Reply:
Mmm. If that's what it happens, I missed it. I'll double check.

Reply:
You were right. I made a modification were the compute function now does not modify the underlying builder.